src/mlpack/methods/ann/layer/linear3d_impl.hpp

```diff
@@ -0,0 +1,178 @@
+/**
+ * @file methods/ann/layer/linear3d_impl.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Implementation of the Linear layer class which accepts 3D input.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_LINEAR3D_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_LINEAR3D_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "linear3d.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType,
+    typename RegularizerType>
+Linear3D<InputDataType, OutputDataType, RegularizerType>::Linear3D() :
+    inSize(0),
+    outSize(0)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType,
+    typename RegularizerType>
+Linear3D<InputDataType, OutputDataType, RegularizerType>::Linear3D(
+    const size_t inSize,
+    const size_t outSize,
+    RegularizerType regularizer) :
+    inSize(inSize),
+    outSize(outSize),
+    regularizer(regularizer)
+{
+  weights.set_size(outSize * inSize + outSize, 1);
+}
+
+template<typename InputDataType, typename OutputDataType,
+    typename RegularizerType>
+void Linear3D<InputDataType, OutputDataType, RegularizerType>::Reset()
+{
+  typedef typename arma::Mat<typename OutputDataType::elem_type> MatType;
+
+  weight = MatType(weights.memptr(), outSize, inSize, false, false);
+  bias = MatType(weights.memptr() + weight.n_elem, outSize, 1, false, false);
+}
+
+template<typename InputDataType, typename OutputDataType,
+    typename RegularizerType>
+template<typename eT>
+void Linear3D<InputDataType, OutputDataType, RegularizerType>::Forward(
+    const arma::Mat<eT>& input, arma::Mat<eT>& output)
+{
+  typedef typename arma::Mat<eT> MatType;
+  typedef typename arma::Cube<eT> CubeType;
+
+  if (input.n_rows % inSize != 0)
+  {
+    Log::Fatal << "Number of features in the input must be divisible by inSize."
+               << std::endl;
+  }
+
+  const size_t nPoints = input.n_rows / inSize;
+  const size_t batchSize = input.n_cols;
+
+  output.set_size(outSize * nPoints, batchSize);
+
+  CubeType inputTemp(const_cast<MatType&>(input).memptr(), inSize, nPoints,
+      batchSize, false, false);
+
+  for (size_t i = 0; i < batchSize; ++i)
+  {
+    // Shape of weight : (outSize, inSize).
+    // Shape of inputTemp : (inSize, nPoints, batchSize).
+    MatType z = weight * inputTemp.slice(i);
+    z.each_col() += bias;
+    output.col(i) = arma::vectorise(z);
```

Comment:
Perhaps, it'll save a memory allocation. However, I'm not sure.
```suggestion
    output.col(i) = arma::vectorise(weight * inputTemp.slice(i) +
        arma::repmat(bias, 1, nPoints));
```

Reply:
But `arma::repmat()` is extending a vector into matrix. Probably this allocates more memory?

Reply:
Armadillo is a tricky library. Sometimes it returns a special object `Operation` instead of matrix. So, the computation itself happens at `operator=`.

Reply:
The documentation says its more efficient to use `each_col()`. https://pasteboard.co/JmOdSSs.png
Also, I tried [this](https://pastebin.com/exVWCxDq) code, which shows a significant time difference. Though I'm still not sure whether space complexity is better or not.

Reply:
Okay, fine.

By the way, there's a slight issue in your benchmark. The compiler could optimize your code out (i.e. remove it) since you don't use the result of the computation. It's easy to fix, just write the result to the screen (or a function of the result like the sum of all elements) after the time measurement.

Reply:
Yes! That made some difference.

But still, I'm getting 0.4 sec difference on my system if I print `arma::accu(output)` in the last line.